### PR TITLE
Add Programmed condition to Gateways

### PR DESCRIPTION
**Pull request by keithmattix**
**Please provide a description of this PR:**

Gateway API is now testing for the Programmed status on Gateways in conformance tests (since kubernetes-sigs/gateway-api#1732). This PR adds the Programmed Status to Gateways (not just listeners within Gateways) and fixes a couple of other Gateway API [conventions](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io%2fv1beta1.GatewayConditionReason).

### CODE REVIEW

**Comment by keithmattix** [comment]
Fixes #43498 43498

File: pilot/pkg/config/kube/gateway/conversion.go [hunk 1]
```diff
@@ -1371,18 +1371,25 @@ func convertGateways(r ConfigContext) ([]config.Config, map[parentKey][]*parentI
 		// Setup initial conditions to the success state. If we encounter errors, we will update this.
 		gatewayConditions := map[string]*condition{
 			string(k8sbeta.GatewayConditionReady): {
-				reason:  "ListenersValid",
+				reason:  string(k8sbeta.GatewayReasonReady),
 				message: "Listeners valid",
 			},
 		}
 		if IsManaged(kgw) {
```
Comment (howardjohn) [inline pilot/pkg/config/kube/gateway/conversion.go:1378]: This is a bit complicated but I think this may be not quite right. The intent here is that if its managed, we set these to negative responses if they are empty (setOnce does the "if empty"); this code doesn't deploy them so it doesn't know if resources are deployed to the cluster yet.

When its NOT managed, we don't have this issue and known all info.


This worked a lot better with Scheduled - the seperation from "we deployed the thing" and "we configured the thing" was explicit in conditions, which mirrored our code.

Its a bit murkier now...

Comment (keithmattix) [reply pilot/pkg/config/kube/gateway/conversion.go:1378]: Yeah I struggled with the exact semantics here a bit. Here was my reasoning:
1. Programmed is a positive-polarity condition so it should [always be set](https://gateway-api.sigs.k8s.io/v1alpha2/references/spec/#gateway.networking.k8s.io%2fv1beta1.GatewayConditionType)
2. If the Gateway IS managed, we can't immediately expect the configuration to propagate "soon" because we can't confirm that the infra is provisioned
3. If the Gateway IS NOT managed, we still haven't generated configuration for it yet (at least based on this function) so I set the error to be true (and only change it at all if the reason is empty or Pending since Pending is the default)
4. If there are invalid listeners, hostnames, etc. that would prevent configuration from becoming available in the Gateway, then we should also set Programmed to False for the "invalid" reason with the message mirroring whatever is preventing the Gateway from moving forward.

In short, if there's anything that threatens the "soon-ness" of configuration being available in the Gateway, then Programmed should be False

Comment (howardjohn) [reply pilot/pkg/config/kube/gateway/conversion.go:1378]: > If the Gateway IS NOT managed, we still haven't generated configuration for it yet (at least based on this function) so I set the error to be true (and only change it at all if the reason is empty or Pending since Pending is the default)

By the time we set the status we have generated the config though, which is why its currently all positive. Throughout the code today we initialize with positive then set negative on errors instead of the inverse. The other approach works too, just not what we had done

Comment (keithmattix) [reply pilot/pkg/config/kube/gateway/conversion.go:1378]: That's fair, and since there's not a single test case that results in the initial error, that strengthens the case that we won't get to this branch. I'll change the initialization to positive
